src/random_bq_to_anndata.py

```diff
@@ -1,46 +1,178 @@
+    # NOTE - really don't need to join to cas_cell_index and cas_feature_index here anymore.
+    sql = f"SELECT matrix.cas_cell_index, original_cell_id, matrix.cas_feature_index, original_feature_id, raw_counts AS count FROM `{project}.{dataset}.cas_cell_info` AS cell, `{project}.{dataset}.cas_feature_info` AS feature, `{project}.{dataset}.cas_raw_count_matrix` AS matrix WHERE matrix.cas_cell_index = cell.cas_cell_index AND matrix.cas_feature_index = feature.cas_feature_index AND" + in_clause + " ORDER BY matrix.cas_cell_index, matrix.cas_feature_index"
     query = client.query(sql)
     return query.result()
 
 
-def random_bq_to_anndata(project, dataset, num_cells):
+def random_bq_to_anndata(project, dataset, num_cells, output_file_prefix):
     client = bigquery.Client(project=project)
-    cell_data = get_cell_data(project, dataset, client, num_cells)
-    adata = ad.AnnData()
-    for row in list(cell_data):
-        # insert VS-507 here
-        pass
+
+    print(f"Getting {num_cells} random cells' data from {project}.{dataset}...")
+    random_cell_ids = get_random_cell_ids(project, dataset, client, num_cells)
+
+    # Read the cell information and store metadata for later
+    ordered_cell_indices, cell_index_to_cell = get_cells(project, dataset, client, random_cell_ids)
+
+    original_cell_ids = []
+    cell_types = []
+    for cell_index in ordered_cell_indices:
+        original_cell_ids.append(cell_index_to_cell[cell_index].original_cell_id)
+        cell_types.append(cell_index_to_cell[cell_index].cell_type)
+
+    # Read the feature information and store metadata for later.
+    ordered_feature_indices, feature_index_to_feature = get_features(project, dataset, client)
+
+    feature_ids = []
+    feature_names = []
+    last_feature_index = None
+    for feature_index in ordered_feature_indices:
+        if (last_feature_index is None):
+            last_feature_index = feature_index - 1
+        # # We expect all entries in cas_feature_info to be stored in continuous fashion. (1, 2, 3, ...)
+        if (feature_index != last_feature_index + 1):
+            raise Exception("ERROR: Non-continuous values for `cas_feature_index` in table `cas_feature_info`")
+
+        feature_ids.append(feature_index_to_feature[feature_index].original_feature_id)
+        feature_names.append(feature_index_to_feature[feature_index].feature_name)
+        last_feature_index = feature_index
+
+    # Note that this method requires that the result set returned by get_cell_data be sorted by cas_cell_index (or, could also be sorted by original_cell_id)
+    cell_data = get_matrix_data(project, dataset, client, random_cell_ids)
+
+    (index_ptr, indices, data) = generate_sparse_matrix(cell_data, ordered_feature_indices[0])
+
+    # Create the matrix from the sparse data representation generated above.
+    counts = csr_matrix((data, indices, index_ptr), dtype=np.float32)
+    adata = ad.AnnData(counts)
+    adata.obs_names = original_cell_ids
```

Comment:
looks like our existing code refers to a similar concept as `adata.obs.index` and `adata.var.index`, is there a difference?

Reply:
Really not sure. I've changed the code to use `adata.obs.index` and `adata.var.index` for consistency. They *seem* to populate the `adata.obs_names` and `adata.var_names` in the anndata structure when I browse it.